### PR TITLE
Add && Impl new trait ToCompactStr

### PR DESCRIPTION
**Pull request by NobodyXu**
Similar to [`std::string::ToString`](https://doc.rust-lang.org/std/string/trait.ToString.html), `ToCompactStr` convert any value implements `fmt::Display` into a `CompactStr` efficiently.

However, due to the fact that `Arc::<[MaybeUninit<T>]>` is still unstable, there is no way to implement `HeapString` efficiently without first creating a `String`, then copy it onto the buffer and deallocate the original `String`.

It seems that the compiler may not be able to optimize these calls, especially with `std::format!` calling `format`.

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

### CODE REVIEW

**Comment by ParkMyCar** [comment]
Thanks for the PR!

I really like the idea of a `ToCompactStr` trait. But you're right, without the nightly compiler we can't do this very efficiently. The `HeapString` representation will be changing soon, see #12, I plan to spend some time on it this weekend. Let's hold off on landing this until we make the changes to `HeapString`

**Comment by NobodyXu** [comment]
Sure, I will wait for #12

**Comment by ParkMyCar** [comment]
Hey @NobodyXu, now that `CompactStr` is more stable, would you want to take another shot at this?

I think the idea of a `ToCompactStr` trait is great, I'm a bit apprehensive of manually implementing it for any type that implements `Display` though because of performance. Maybe we could define the trait, and then manually implement for a few common types like all the number types (e.g. `u8`, `isize`, `f32`, etc), and then maybe `bool` and `char`? IMO that would go a fairly long way for common use cases, what do you think?

For efficiently converting number types to strings take a look at [`itoa`](https://github.com/dtolnay/itoa) and [`ryu`](https://github.com/dtolnay/ryu)

**Comment by NobodyXu** [comment]
> Hey @NobodyXu, now that `CompactStr` is more stable, would you want to take another shot at this?

Yes, sure.

> I think the idea of a `ToCompactStr` trait is great, I'm a bit apprehensive of manually implementing it for any type that implements `Display` though because of performance. Maybe we could define the trait, and then manually implement for a few common types like all the number types (e.g. `u8`, `isize`, `f32`, etc), and then maybe `bool` and `char`? IMO that would go a fairly long way for common use cases, what do you think?

Hmmm, I don't like the idea of implementing only for common types, as it would make `ToCompactStr` less useful and
provide different semantics compared to `ToString`.

For common number types, maybe we can provide a optional specialization for them using
 [castaway](https://crates.io/crates/castaway) and gate it using a feature flag?

**Comment by ParkMyCar** [comment]
> Hmmm, I don't like the idea of implementing only for common types, as it would make ToCompactStr less useful and
provide different semantics compared to ToString.
Fair enough, that makes sense to me!

Ugh I wish [specialization](https://rust-lang.github.io/rfcs/1210-impl-specialization.html) was stable. `castaway` looks pretty cool, let's give it a shot!

**Comment by NobodyXu** [comment]
> > Hmmm, I don't like the idea of implementing only for common types, as it would make ToCompactStr less useful and
> > provide different semantics compared to ToString.
> 
> Fair enough, that makes sense to me!
> 
> Ugh I wish [specialization](https://rust-lang.github.io/rfcs/1210-impl-specialization.html) was stable. Regardless, `castaway` looks pretty cool, let's give it a shot!

I am quite busy recently, so it might take me some time to implement it.

**Comment by NobodyXu** [comment]
Sorry I accidentally closed this PR, now it is reopened.

**Comment by NobodyXu** [comment]
@ParkMyCar I have implemented trait `ToCompactStr` without specialisation using [castaway](https://docs.rs/castaway/0.1.2/castaway/macro.cast.html).

I will implement the specialisation later, but now you can probably start looking at my code and providing me with feedback so that we can ensure the implementation is correct before doing any optimization.

File: compact_str/src/lib.rs
Comment (ParkMyCar) [inline compact_str/src/lib.rs:19]: Can you re-run `cargo +nightly fmt`? We use some rules which are nightly only

File: compact_str/src/lib.rs
Comment (ParkMyCar) [inline compact_str/src/lib.rs:705]: This example doesn't actually use `.to_compact_str()`?

File: compact_str/src/lib.rs
Comment (ParkMyCar) [inline compact_str/src/lib.rs:722]: `CompactStr` now implements `fmt::Write`, so I think this entire impl could just be
```
fn to_compact_str(&self) -> CompactStr {
  let mut compact = CompactStr::new("");
  write!(compact, "{}", self).expect("fmt::Display incorrectly implemented!");
  compact
}
```

File: compact_str/src/utility.rs
Comment (ParkMyCar) [inline compact_str/src/utility.rs:13]: This is pretty neat! See above though, I don't think we need to do this since `CompactStr` impls `fmt::Write`

**Review by ParkMyCar — CHANGES_REQUESTED** [review]
Thanks for doing this work!

File: compact_str/src/lib.rs
Comment (NobodyXu) [reply compact_str/src/lib.rs:705]: Oops, I simply copied it from the old PR diff.
I will change that.

File: compact_str/src/lib.rs
Comment (NobodyXu) [reply compact_str/src/lib.rs:722]: That would make the code much shorter.

However, I think we should benchmark the code first, as the current implementation might provide better performance than using fmt::Write, since everything is written in one-go.

For size <= MAX_SIZE, current implementation can be optimized into one copy of array, or no copy at all.

For size > MAX_SIZE, current implementation would only allocate once, where fmt::Write could allocate multiple times.

Comment (NobodyXu) [reply compact_str/src/lib.rs:722]: I have pushed a new commit that optimize the case where size > `MAX_SIZE` and preallocate the `String` with the exact size and use `fmt::Write` to write the format string.

This would avoid the repeated allocation that occurs if using `fmt::Write` implementation of `CompactStr`.

File: compact_str/src/lib.rs
Comment (NobodyXu) [reply compact_str/src/lib.rs]: Done.

File: compact_str/src/lib.rs
Comment (NobodyXu) [reply compact_str/src/lib.rs:19]: That explains why the CI complains about the code style, while my neovim is configured to automatically format my code using `cargo fmt`.

I will run `cargo +nightly fmt` just before merging so that I don't have to run it repeatively.

File: compact_str/src/repr/mod.rs
Comment (ParkMyCar) [inline compact_str/src/repr/mod.rs:120]: We should either `impl Write` for `Repr` or have this method return a `CompactStr`, for two reasons:
1. Writing into a buffer and then using that buffer to create a `Repr` results in an additional copy
2. Using something like `format_into_buffer(...)` is another part of code that manually writing bytes and ideally we have as few of those as possible, since these are the easiest things to get wrong

Comment (ParkMyCar) [inline compact_str/src/repr/mod.rs:120]: We should either `impl Write` for `Repr` or have this method return a `CompactStr`, for two reasons:
1. Writing into a buffer and then using that buffer to create a `Repr` results in an additional copy
2. Using something like `format_into_buffer(...)` is another part of code that manually writing bytes and ideally we have as few of those as possible, since these are the easiest things to get wrong

File: compact_str/src/lib.rs
Comment (ParkMyCar) [reply compact_str/src/lib.rs:724]: I ran some benchmarks with this implementation and with just using `write!(...)`. In the simple case (e.g. `&str`, `usize`, etc.) just using `write!(...)` is a lot faster, especially for short strings. It's when we have long strings with multiple arguments (e.g. a struct) that counting all of the bytes beforehand and pre-allocating saves a lot of time.

Given this, the approach of initially counting bytes works for me, with the idea that we can optimize for the simple cases later on.

Comment (ParkMyCar) [reply compact_str/src/lib.rs:724]: I ran some benchmarks with this implementation and with just using `write!(...)`. In the simple case (e.g. `&str`, `usize`, etc.) just using `write!(...)` is a lot faster, especially for short strings. It's when we have long strings with multiple arguments (e.g. a struct) that counting all of the bytes beforehand and pre-allocating saves a lot of time.

Given this, the approach of initially counting bytes works for me, with the idea that we can optimize for the simple cases later on.

File: compact_str/src/repr/mod.rs
Comment (NobodyXu) [reply compact_str/src/repr/mod.rs]: I have fixed this performance issue by implementing `InlineString::from_fmt` and call the `format_into_buffer` directly on the buffer used by `InlineString`.

Comment (NobodyXu) [reply compact_str/src/repr/mod.rs]: I have now refactored the code and inline `format_into_buffer` into `InlineString::from_fmt`.

Since `format_into_buffer` simply uses `std::io::Cursor`, it's unlikely to have any bugs.

Comment (NobodyXu) [reply compact_str/src/repr/mod.rs]: I have pushed another commit to simplify `InlineString::from_fmt`, by removing use of `std::io::Cursor`.

**Comment by NobodyXu** [comment]
> For efficiently converting number types to strings take a look at [`itoa`](https://github.com/dtolnay/itoa) and [`ryu`](https://github.com/dtolnay/ryu)

I am doubtful on using [`ryu`](https://github.com/dtolnay/ryu) as it might give different output compared to the builtin `format!`, I think this might break some code.

**Comment by NobodyXu** [comment]
> > Hmmm, I don't like the idea of implementing only for common types, as it would make ToCompactStr less useful and
> > provide different semantics compared to ToString.
> 
> Fair enough, that makes sense to me!
> 
> Ugh I wish [specialization](https://rust-lang.github.io/rfcs/1210-impl-specialization.html) was stable. Regardless, `castaway` looks pretty cool, let's give it a shot!

I just realized that `castaway` requires the type to be of `'static` lifetime, which is impossible to use within `ToCompactStr`.

I am now thinking of adding another trait `ToCompactStrFast` for the builtin type.

**Comment by NobodyXu** [comment]
I also asked the question https://github.com/sagebind/castaway/issues/5 on `castaway` to see if there is any alternative solution.

**Comment by NobodyXu** [comment]
> > > Hmmm, I don't like the idea of implementing only for common types, as it would make ToCompactStr less useful and
> > > provide different semantics compared to ToString.
> > 
> > 
> > Fair enough, that makes sense to me!
> > Ugh I wish [specialization](https://rust-lang.github.io/rfcs/1210-impl-specialization.html) was stable. Regardless, `castaway` looks pretty cool, let's give it a shot!
> 
> I just realized that `castaway` requires the type to be of `'static` lifetime due to `std::any::TypeId::of` requires static lifetime, which is impossible to use within `ToCompactStr`.
> 
> I am now thinking of adding another trait `FromBuiltinTypes`.
> 
> It is just unfortunate that Rust doesn't have specialisation stabilised.

With https://github.com/sagebind/castaway/pull/6 merged in and released as https://github.com/sagebind/castaway/releases/tag/0.2.0, this is no longer an issue.

**Comment by NobodyXu** [comment]
@ParkMyCar I've added int and float specialisation, but there are two problems:
 - [ ] First is `castaway::LifetimeFree` is not implemented for `NonZeroUsize` so that we cannot provide specialisation for it yet https://github.com/sagebind/castaway/pull/7
 - [ ] Second is that test on float specialisation (line 421 of `tests.rs`) fails on power pc little endian 64-bit:
        Expected: 3.2
        Actual: 0.0
        It failed only after float specialisation is added.

File: compact_str/src/to_compact_str_specialisation.rs
Comment (sagebind) [inline compact_str/src/to_compact_str_specialisation.rs:28]: Hey, just snooping on this PR as I like to see how Castaway is being used! I'll double-check to make sure that this bug isn't coming from Castaway's side.

Comment (NobodyXu) [reply compact_str/src/to_compact_str_specialisation.rs:28]: Awesome!

Since `castaway` is working as intended, I think the bug is more likely to be from the float formatting algorithm in std and ryu.

Seems like they had a hard time running on `powerpc64le` target.

Comment (sagebind) [reply compact_str/src/to_compact_str_specialisation.rs:31]: Yeah I figured that was the case, but I am beefing up my tests just to be sure. I'm confident in Castaway's logic, but when internally I have a `transmute_unchecked` function because the already-dangerous builtin `transmute` is _too strict_ you can't be too careful. 😀

**Comment by NobodyXu** [comment]
I just tried to reproduce this on my own x86_64 linux box by running:

```
cross +nightly test --release --all-features --manifest-path=compact_str/Cargo.toml --target powerpc-unknown-linux-gnu
```

however, all tests passed without problem...

File: compact_str/src/to_compact_str_specialisation.rs
Comment (NobodyXu) [reply compact_str/src/to_compact_str_specialisation.rs]: This is confirmed to be a bug in rust as I can reproduce this on my own x86_64 linux box using `cross`, so would have to wait for rust to fix it.

File: .github/workflows/msrv.yml
Comment (ParkMyCar) [inline .github/workflows/msrv.yml:28]: Why did you get rid of `cargo hack`? It's quite useful in making sure `compact_str` works for all versions of rustc since the MSRV, and all different combination of features?

File: compact_str/Cargo.toml
Comment (ParkMyCar) [inline compact_str/Cargo.toml:17]: I would really prefer to not have so many features, especially ones tied to performance, can we make these dependencies required instead of optional?

File: compact_str/src/repr/inline.rs
Comment (ParkMyCar) [inline compact_str/src/repr/inline.rs:52]: This introduces another part of the code base that directly writes into the buffer and needs to know about the underlying bit layout, I'd like to prevent that. 

See my comment on `impl<T: fmt::Display> ToCompactStr for T` on how we can eliminate this

File: compact_str/src/lib.rs
Comment (ParkMyCar) [inline compact_str/src/lib.rs:779]: Can we implement `fmt::Write` for `Repr`, and then here do something like:
```
let num_bytes = count(&self);
let mut repr = Repr::with_capacity(num_bytes);
write!(&mut repr, "{}", self).expect(...);
```

File: compact_str/src/tests.rs
Comment (ParkMyCar) [inline compact_str/src/tests.rs:381]: Is the purpose of this macro to be similar to `format!(...)`? If so, could we rename this to `format_compact!(...)`

File: compact_str/src/to_compact_str_specialisation.rs
Comment (ParkMyCar) [inline compact_str/src/to_compact_str_specialisation.rs:16]: I'll need to take a closer look at this to figure out how it's actually zero cost, right now there seem to be some runtime checks?

Out of curiosity, is it not possible to do autoref specialization here?

**Review by ParkMyCar — CHANGES_REQUESTED** [review]
Thanks for all the work here @NobodyXu! I've been pretty busy lately so haven't had a chance to look super in-depth, but I think we're getting close!

File: compact_str/src/to_compact_str_specialisation.rs
Comment (NobodyXu) [reply compact_str/src/to_compact_str_specialisation.rs:16]: According to @sagebind , all of these cast! can be optimized out by compiler.

I briefly read through the code and I also think it is possible for compiler to prune all the unused specialisation.

File: compact_str/Cargo.toml
Comment (NobodyXu) [reply compact_str/Cargo.toml:17]: sure

File: .github/workflows/msrv.yml
Comment (NobodyXu) [reply .github/workflows/msrv.yml:28]: They are taking too long to run, probably because of the new features I added

File: compact_str/src/to_compact_str_specialisation.rs
Comment (sagebind) [reply compact_str/src/to_compact_str_specialisation.rs:16]: "runtime checks" is a bit hard to define in light of optimization; technically yes, Castaway macros generate program code to make appropriate checks. But these checks are trivial from the compiler's perspective (think `if size_of::<A>() == size_of::<B>()`) and are always reduced to a simple `true` or `false` at even just a basic optimization level; once this happens it is also trivial for the compiler to eliminate a branch in _your_ code, since it effectively looks like `if true {} else {}` (or `if false {} else {}` depending on which type it is called with) to rustc by the time that stage of compilation is reached.

> Out of curiosity, is it not possible to do autoref specialization here?

Nope, auto(de)ref techniques don't work inside a generic context, because ref specialization is resolved _before_ monomorphization happens. (Which is a good thing by the way, otherwise you'd have C++ problems, such as compilation failing "sometimes" depending on which concrete types you supply to a template.)

File: compact_str/Cargo.toml
Comment (NobodyXu) [reply compact_str/Cargo.toml]: Done!

File: compact_str/src/tests.rs
Comment (NobodyXu) [reply compact_str/src/tests.rs]: Done!

File: compact_str/src/repr/inline.rs
Comment (NobodyXu) [reply compact_str/src/repr/inline.rs]: Done!

File: compact_str/src/lib.rs
Comment (NobodyXu) [reply compact_str/src/lib.rs]: Done!

I've taken things even further, by using `CompactStr::with_capacity` then use `fmt::Write` implementation for `CompactStr` to format `self` into `compact_str.`

File: .github/workflows/msrv.yml
Comment (NobodyXu) [reply .github/workflows/msrv.yml]: OK, I've reverted the changes to this workflow.

File: compact_str/src/lib.rs
Comment (NobodyXu) [reply compact_str/src/lib.rs]: I've applied `cargo +nightly fmt --all`.

SInce this PR is pretty close to being merged, I would apply `cargo +nightly fmt --all` everytime I push new commits.

File: compact_str/src/to_compact_str_specialisation.rs
Comment (NobodyXu) [reply compact_str/src/to_compact_str_specialisation.rs:19]: @ParkMyCar [Here is the code](https://github.com/sagebind/castaway/blob/52987d6ebbdfcc451fb24585dcb34590f6e62657/src/utils.rs#L29) responsible for checking whether two non-`' static` types are the same, which is at the core of `castaway`.

I am pretty sure they can be easily evaluated at compile time.

Comment (ParkMyCar) [reply compact_str/src/to_compact_str_specialisation.rs:19]: > since it effectively looks like `if true {} else {}` (or `if false {} else {}`)
That's wild, super cool work @sagebind!

**Review by ParkMyCar — APPROVED** [review]
Thanks for all the hard work here @NobodyXu and making all the changes I asked for! 

Looks like there's just one failing test on PowerPC Little Endian with float formatting? If we can fix the test that would be great, otherwise I'm fine with ignoring it. Once you make that change we can merge!

Thanks again for all the work! 🙂

**Comment by NobodyXu** [comment]
> Looks like there's just one failing test on PowerPC Little Endian with float formatting? If we can fix the test that would be great, otherwise I'm fine with ignoring it. Once you make that change we can merge!

Unfortunately that isna compiler bug https://github.com/rust-lang/rust/issues/96306 which I cannot fix.

**Comment by ParkMyCar** [comment]
Merging, thanks @NobodyXu!

**Comment by NobodyXu** [comment]
@ParkMyCar Now that we have more unit tests and fuzzing added for `ToCompactStr`, maybe it is a good time to release a new minor version?

**Pull request merged**
